### ISSUE TITLE
Using NULL and NOT NULL in queries

### ISSUE THREAD

**Issue by atyndall** [Dynamoid/dynamoid#358]
I noticed that the use of `thing.null` and `thing.not_null` filters are not supported by the library, as mentioned in [aws_sdk_v3.rb](https://github.com/Dynamoid/dynamoid/blob/master/lib/dynamoid/adapter_plugin/aws_sdk_v3.rb#L25):

>Don't implement NULL and NOT_NULL because it doesn't make seanse - we declare schema in models

I find myself in the position where being able to filter results based on the null/not_null-ness of an attribute would be very useful. Is there a way to accomplish this with the library? If not, could you elaborate on the reasons why it isn't implemented?

Much appreciated!

**Comment by andrykonchin** [comment]
According to the [documentation](https://docs.aws.amazon.com/en_us/amazondynamodb/latest/APIReference/API_Condition.html) `NULL` and `NOT_NULL` predicates mean presence of some attribute in an item/document and not just compare attribute's value with `NULL`.

As far as Dynamoid has some kind of schema items created with it always have all declared attributes but some with `NULL` values. So `NULL` and `NOT_NULL` predicates may not make sense - each attribute will be `NOT_NULL`. 

I can imagine situation when these predicates may be useful - when a schema was changed and some new fields/attributes were added...

Actually it isn't a big deal to add these predicates. The only intention to not to implement them was to not confuse the users because of this special semantic of the predicates.

So could you please describe your use case?

**Comment by atyndall** [comment]
Hi Andrew, thanks for getting back to me.

This is where things get a bit confusing for me, because when I use the Dynamoid library, I am not observing the `NULL` behaviour you describe. When I set a `nil` value for an attribute, Dynamoid does not set the attribute at all, as opposed to setting it with a `NULL` value.

A simple test case;

```ruby
class Dynamo::Foo
  include Dynamoid::Document

  table name: :my_table
  range :sort, :string
  field :id, :string
  field :created_at, :datetime
  field :version, :string
  field :bar, :number
end
```

When I insert, you can see that the `version` attribute is completely unset;
```
2.5.1 :014 > Dynamo::Foo.create!(id: 'test-1', sort: 'foo', created_at: Time.now, version: nil, bar: 1)

put_item | Request "{\"TableName\":\"my_table\",\"Item\":{\"created_at\":{\"N\":\"1559261279.659766529\"},\"updated_at\":{\"N\":\"1559261279.660421395\"},\"id\":{\"S\":\"test-1\"},\"sort\":{\"S\":\"foo\"},\"bar\":{\"N\":\"1.0\"}},\"Expected\":{\"id\":{\"Exists\":false},\"sort\":{\"Exists\":false}}}" | Response "{}"
(32.32 ms) PUT ITEM - ["my_table", {:created_at=>0.1559261279659766529e10, :updated_at=>0.1559261279660421395e10, :id=>"test-1", :sort=>"foo", :version=>nil, :bar=>0.1e1}, {}]

#<Dynamo::Foo:0x000055fc66faf498 @attributes={:created_at=>Fri, 31 May 2019 00:07:59 +0000, :updated_at=>Fri, 31 May 2019 00:07:59 +0000, :id=>"test-1", :sort=>"foo", :version=>nil, :bar=>0.1e1}, @new_record=false, @associations={}, @attributes_before_type_cast={:created_at=>2019-05-31 00:07:59 +0000, :updated_at=>Fri, 31 May 2019 00:07:59 UTC +00:00, :id=>"test-1", :sort=>"foo", :version=>nil, :bar=>1}, @attributes_changed_by_setter={}, @mutations_from_database=nil, @validation_context=nil, @errors=#<ActiveModel::Errors:0x000055fc66facf68 @base=#<Dynamo::Foo:0x000055fc66faf498 ...>, @messages={}, @details={}>, @previously_changed={}, @mutations_before_last_save=nil>
```

viewed within the DynamoDB console;

<img width="512" alt="Screen Shot 2019-05-31 at 8 10 49 am" src="https://user-images.githubusercontent.com/1833303/58672916-ab51ed80-837b-11e9-9df2-1d729976cf5c.png">

and when I insert with the `version` attribute set, it sets it;

```
2.5.1 :016 > Dynamo::Foo.create!(id: 'test-1', sort: 'foo2', created_at: Time.now, version: 'wow!', bar: 1)

put_item | Request "{\"TableName\":\"my_table\",\"Item\":{\"created_at\":{\"N\":\"1559261388.5800343\"},\"updated_at\":{\"N\":\"1559261388.580517919\"},\"id\":{\"S\":\"test-1\"},\"sort\":{\"S\":\"foo2\"},\"version\":{\"S\":\"wow!\"},\"bar\":{\"N\":\"1.0\"}},\"Expected\":{\"id\":{\"Exists\":false},\"sort\":{\"Exists\":false}}}" | Response "{}"
(27.56 ms) PUT ITEM - ["my_table", {:created_at=>0.15592613885800343e10, :updated_at=>0.1559261388580517919e10, :id=>"test-1", :sort=>"foo2", :version=>"wow!", :bar=>0.1e1}, {}]

#<Dynamo::Foo:0x000055fc6771b308 @attributes={:created_at=>Fri, 31 May 2019 00:09:48 +0000, :updated_at=>Fri, 31 May 2019 00:09:48 +0000, :id=>"test-1", :sort=>"foo2", :version=>"wow!", :bar=>0.1e1}, @new_record=false, @associations={}, @attributes_before_type_cast={:created_at=>2019-05-31 00:09:48 +0000, :updated_at=>Fri, 31 May 2019 00:09:48 UTC +00:00, :id=>"test-1", :sort=>"foo2", :version=>"wow!", :bar=>1}, @attributes_changed_by_setter={}, @mutations_from_database=nil, @validation_context=nil, @errors=#<ActiveModel::Errors:0x000055fc6771a160 @base=#<Dynamo::Foo:0x000055fc6771b308 ...>, @messages={}, @details={}>, @previously_changed={}, @mutations_before_last_save=nil>
```

in the console;

<img width="493" alt="Screen Shot 2019-05-31 at 8 10 57 am" src="https://user-images.githubusercontent.com/1833303/58672926-b1e06500-837b-11e9-8ee4-be74d90adbb9.png">

So this is all a bit strange, as the implementation changed somehow from how you expect?

But the practical upshot is that using the Dynamo `NULL`/`NOT_NULL` filter is actually meaningful for my use case as it can disambiguate the `version` attribute being set or not set.

**Comment by andrykonchin** [comment]
Hey

I added new operators (https://github.com/Dynamoid/dynamoid/pull/359) so it should solve your issue.

Also I added new config option `store_attribute_with_nil_value` to make Dynamoid store attributes with `nil` values in DynamoDB item (https://github.com/Dynamoid/dynamoid/pull/360). So if it's OK to recreate the DynamoDB table it makes more sense to set `true` value and use ordinal way to query:

```ruby
Model.where(versions: nil)
Model.where('versions.ne': nil)
```

**Comment by atyndall** [comment]
Thank you, this is very helpful! I'll await a new release and give it a try.

**Comment by andrykonchin** [comment]
The changes were released in 3.3.0.

Please let me know if there is any problem. I will close the issue until we can confirm it doesn't work well.
